cmake/cpu_features.cmake

```diff
@@ -127,6 +131,57 @@ else ()
     if (HAVE_AVX2 AND ENABLE_AVX2)
         set (COMPILER_FLAGS "${COMPILER_FLAGS} ${TEST_FLAG}")
     endif ()
+
+    set (TEST_FLAG "-mavx512f -mavx512bw")
+    set (CMAKE_REQUIRED_FLAGS "${TEST_FLAG} -O0")
+    check_cxx_source_compiles("
+        #include <immintrin.h>
+        int main() {
+            auto a = _mm512_setzero_epi32();
+            (void)a;            
+            auto b = _mm512_add_epi16(__m512i(), __m512i());
+            (void)b;
+            return 0;
+        }
+    " HAVE_AVX512)
+    if (HAVE_AVX512 AND ENABLE_AVX512)
+        set (COMPILER_FLAGS "${COMPILER_FLAGS} ${TEST_FLAG}")
+    endif ()
+
+    set (TEST_FLAG "-mbmi")
+    set (CMAKE_REQUIRED_FLAGS "${TEST_FLAG} -O0")
+    check_cxx_source_compiles("
+        #include <immintrin.h>
+        int main() {
+            auto a = _blsr_u32(0);
+            (void)a;
+            return 0;
+        }
+    " HAVE_BMI)
+    if (HAVE_BMI AND ENABLE_BMI)
+        set (COMPILER_FLAGS "${COMPILER_FLAGS} ${TEST_FLAG}")
+    endif ()   
+
+#Limit avx2/avx512 flag for specific source build
+    set (X86_INTRINSICS_FLAGS "")
+    if (ENABLE_AVX2_FOR_SPEC_OP)
```

Comment:
What's the motivation of this code?

Reply:
ENABLE_AVX2/ENABLE_AVX512 means compile with AVX2/512 globally enabled.
ENABLE_AVX2_FOR_SPEC_OP/ENABLE_AVX512_FOR_SPEC_OP means enable AVX2/AVX512 when compiling some specific source file.

Reply:
Yeah. The purpose of these options is very clear. But I don't understand the benefit to only enable AVX2/AVX512 for certain source files instead of enabling it globally.

Reply:
Because the performance of AVX-512 may vary on different situations. But for certain source, we can make sure it works as expected.